docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,176 @@
+"How much room do I have in my cluster?" is a very predictable question that users importing huge files will want to know.  We should include cluster cli tools (something like ipfs metrics, or the still-unmade connectivity-graph) that inform users how much space they have available, and the largest recommended file the cluster can ingest.  Perhaps later on we can get even fancier and include predictions of insertion time, and a visualization of how many shards of what size will be added.
+
+Although the ipfs add proxy endpoint is the logical place to launch the adding process we could also provide a user friendly command that takes in the option of a network address to add from.  This way cluster could properly manage the connection with a remote server while buffering and potentially blocking for long periods of time and prevent the user from having to do clever unix things to download large files.  Note I don't have much experience with this kind of thing, so maybe wget etc have really good UX for things like this and this would be overkill.
+
+### Challenges adding files
+When files become larger than local repos this becomes challenging for two reasons.  First, the importing node now needs to be able to block the importing process periodically to transfer data over to the pinning peer so that its local resources (ipfs repo and maybe memory) don't get overloaded.  Many related complications also seem likely.  For example if one shard takes up 90% of the node's ipfs repo and if the importing node makes use of its repo for temporary pinning and transfer then the importing node cannot pin its own shard until all other nodes have been served.  Second, I am unclear if either of the two proposed data transfer methods will provide acceptable performance when transferring large amounts of data (e.g. on the order of 10 GB shards).
+
+The first challenge will require rethinking the go-ipfs dag building functionality and on the bright side may contribute to building a versatile foundation for the potential dagger/dex project.  The second challenge is best addressed by evaluating the performance characteristics of both data transfer methods.  Perhaps the ongoing ipfs metrics work will already answer many questions about the first method.  If the second transfer method (libp2p rpc) is a serious contender we should plan on evaluating its performance as part of implementation.  (TODO -- create an issue defining some metrics that we will want to evaluate) (TODO -- see if any existing ipfs metrics are relavent to first method, e.g. measuring remote pin time of dags with many large leaves)
+
+### Collaborative importing
+We could potentially use alternatives to importing the stream on one master node that then sends out shards to the assigned pinning nodes.  One other possibility is "collaborative importing", where the cluster node doing the "add" operation streams the raw data out to the pinning nodes that then do importing and pinning in place.
```

Comment:
This is an interesting idea. I don't know how much improvement there would be between sending 'chunks' to other peers and 'sending a stream to chunk', and how difficult it would be and how much performance would be added. It seems simpler to chunk everything in one place.

Reply:
Agree on all points especially on the difficulty implementing.  I'll take this out for now.  We can reconsider approaches like this if importing becomes a bottleneck on data streaming to an extent that specific use cases are suffering.